cpp/src/fil/internal.cuh

```diff
@@ -89,19 +90,22 @@ enum output_t {
 };
 
 /** val_t is the payload within a FIL leaf */
+template <typename real_t>
 union val_t {
```

Comment:
Not a requirement, but possibly `std::variant` for type safety?

Reply:
Unfortunately, this is not possible: doing so would increase the size of `val_t`, increase the size of FIL tree nodes beyond the 8-16 bytes they currently are, and would negatively affect FIL performance.